src/Tools/Extensions.ApiDescription.Client/src/build/Microsoft.Extensions.ApiDescription.Client.targets

```diff
@@ -100,44 +101,57 @@
       <!-- If OutputPath is a file, add it directly to relevant items. -->
       <TypeScriptCompile Include="@(_Files)"
           Exclude="@(TypeScriptCompile)"
-          Condition="'%(_Files.OutputPathExtension)' == '.ts' OR '%(_Files.OutputPathExtension)' == '.tsx'">
-        <SourceDocument>%(_Files.FullPath)</SourceDocument>
-      </TypeScriptCompile>
+          Condition="'%(_Files.Extension)' == '.ts' OR '%(_Files.Extension)' == '.tsx'" />
 
       <Compile Include="@(_Files)"
           Exclude="@(Compile)"
-          Condition="'$(DefaultLanguageSourceExtension)' != '.ts' AND '%(_Files.OutputPathExtension)' == '$(DefaultLanguageSourceExtension)'">
-        <SourceDocument>%(OpenApiReference.FullPath)</SourceDocument>
-      </Compile>
+          Condition="'$(DefaultLanguageSourceExtension)' != '.ts' AND '%(_Files.Extension)' == '$(DefaultLanguageSourceExtension)'" />
+    </ItemGroup>
 
-      <!-- Otherwise, add all descendant files with the expected extension. -->
-      <TypeScriptCompile Include="@(_Directories -> '%(Identity)/**/*.ts;%(Identity)/**/*.tsx')"
-          Exclude="@(TypeScriptCompile)">
-        <SourceDocument>%(_Directories.FullPath)</SourceDocument>
-      </TypeScriptCompile>
+    <!--
+      Otherwise, add all descendant files with the expected extension. Glob into _Directories before updating
+      TypeScriptCompile or Compile items. See <https://github.com/dotnet/msbuild/issues/3274> and workaround in
+      <https://stackoverflow.com/questions/48868060/can-a-task-itemgroup-glob-files>. Unfortunately, this workaround
+      loses SourceDocument and other metadata.
+    -->
+    <PropertyGroup>
+      <_TypeScriptCompileItemsFromDirectories>@(_Directories -> '%(Identity)/**/*.ts;%(Identity)/**/*.tsx')</_TypeScriptCompileItemsFromDirectories>
+      <_CompileItemsFromDirectories>@(_Directories -> '%(Identity)/**/*$(DefaultLanguageSourceExtension)')</_CompileItemsFromDirectories>
+    </PropertyGroup>
 
-      <Compile Include="@(_Directories -> '%(Identity)/**/*.$(DefaultLanguageSourceExtension)')"
+    <ItemGroup>
+      <TypeScriptCompile Include="$(_TypeScriptCompileItemsFromDirectories)" Exclude="@(TypeScriptCompile)" />
+
+      <Compile Include="$(_CompileItemsFromDirectories)"
           Exclude="@(Compile)"
-          Condition="'$(DefaultLanguageSourceExtension)' != '.ts'">
-        <SourceDocument>%(_Directories.FullPath)</SourceDocument>
-      </Compile>
+          Condition="'$(DefaultLanguageSourceExtension)' != '.ts'" />
 
-      <FileWrites Exclude="@(FileWrites)"
-          Include="@(_Files);@(_Directories -> '%(Identity)/**/*.ts;%(Identity)/**/*.tsx;%(Identity)/**/*.$(DefaultLanguageSourceExtension)')" />
+      <FileWrites Include="@(_Files);$(_TypeScriptCompileItemsFromDirectories);$(_CompileItemsFromDirectories)"
+          Exclude="@(FileWrites)" />
 
       <_Files Remove="@(_Files)" />
       <_Directories Remove="@(_Directories)" />
     </ItemGroup>
+
+    <PropertyGroup>
```

Comment:
Probably worth one more comment here explaining why you empty these out

Reply:
About the only comment would be something like "Clean up local properties." That is, I just tend to clean up items and properties as I go, without a specific reason. Doing this _might_ give `msbuild` some memory back but I doubt anything measurable. Let me know if you feel strongly about this or have a specific suggestion on wording.

/fyi the `@(_Files)` cleanup at [the top of the target](https://github.com/dotnet/aspnetcore/pull/39146/files#diff-39edc29cf7321e223b9cf94de7f369c369f42db41a220fe164d5149f8427dc5aR102) is much more important because we have other temporary `@(_Files)` item groups.

Reply:
I won't block on it